### PR TITLE
Separate nbnxm setup from the user code

### PR DESCRIPTION
**Pull request by kanduri**
I managed to separate the user facing client code from the underlying GMX code using a translation layer.

The ForceCalculator object would contain in it two other objects. One of them would be the `GmxForceCalculator` which would work with all GMX data structures internally and the other would be a builder (`NbvSetupUtil`) for this that would take in the data structures created by the user and fill the required GMX structures underneath.

I also attempted to separate the data created in the beginning of the simulation and the data transformed in each step.

I am yet to upload tests. I am still writing tests and modifying code to make it more testable. In the process, I am splitting the functions a bit more so that I can limit their scope for better unit testing.

### CODE REVIEW

**Comment by kanduri** [comment]
Some observations about the design:

- `GMXForceCalculator` knows nothing about the user level data structures and the simulation state. It is meant to evolve for the singular purpose of orchestrating internals of GROMACS.

- `NbvSetupUtil` (tentative naming, please suggest others) is the translation layer that holds a handle to the user level data structures and sets up the `GMXForceCalculator`

- Both of the above objects are structs with fully public interfaces and attributes. However, they are private members of `ForceCalculator`. User trying to script particle simulations cannot (and should not) be able to access it. However, we would be able to **test the functionality without using the client code**.

Looking brainstorm a bit on the tests we need to write and how we should invoke them.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+            real c12_combo = detail::combinationFunction(c12_1, c12_2, CombinationRule::Geometric);
+            nonbondedParameters_.push_back(c6_combo);
+            nonbondedParameters_.push_back(c12_combo);
+        }
+    }
+
+    atomInfoAllVdw_.resize(numAtoms);
+    for (size_t atomI = 0; atomI < numAtoms; atomI++)
+    {
+        SET_CGINFO_HAS_VDW(atomInfoAllVdw_[atomI]);
```
Comment (victorusu) [inline src/gromacs/nblib/gmxsetup.cpp:260]: You still need to add `SET_CGINFO_HAS_Q`

**Review by victorusu — CHANGES_REQUESTED** [review]

File: src/gromacs/nblib/gmxsetup.cpp [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+#include "gromacs/utility/logger.h"
+#include "gromacs/utility/smalloc.h"
+
+namespace nblib
+{
+
+namespace detail
+{
+
+static real combinationFunction(real v, real w, CombinationRule combinationRule)
```
Comment (victorusu) [inline src/gromacs/nblib/gmxsetup.cpp:78]: I think this function should be named differently, it has to be a verb also... 
Something like `generateFromCombinationRule(...`, but I am not sure about the name...

**Review by victorusu — CHANGES_REQUESTED** [review]

File: src/gromacs/nblib/atomtype.cpp [hunk 3]
```diff
@@ -49,9 +49,9 @@
 namespace nblib
 {
 
-AtomType::AtomType() noexcept : name_(AtomName("")), mass_(Mass(0)), c6_(C6(0)), c12_(C12(0)) {}
+AtomType::AtomType() noexcept : name_(AtomName("")), mass_(Mass(0)), c6_(C6Param(0)), c12_(C12Param(0)) {}
```
Comment (victorusu) [inline src/gromacs/nblib/atomtype.cpp]: I think we should not use `C6Param` and `C12Param` but only `C6` and `C12`

**Review by victorusu — CHANGES_REQUESTED** [review]

File: src/gromacs/nblib/atomtype.cpp (hunk 3 above)
Comment (kanduri) [reply src/gromacs/nblib/atomtype.cpp]: Yes. This was due to a conflict happening with a namesake macro defined in GROMACS.

**Review by sekelle — APPROVED** [review]

**Pull request merged**